File: packages/ios/assets/ios-xctest-agent/AsturAgentServer.swift
import Foundation
import Network

final class AsturAgentServer {
    private let port: UInt16
    private let agent: AsturAgent
    private let queue = DispatchQueue(label: "dev.astur.ios-agent.server")
    private var listener: NWListener?

    init(port: UInt16, agent: AsturAgent) {
        self.port = port
        self.agent = agent
    }

    func start() throws {
        let listener = try NWListener(using: .tcp, on: NWEndpoint.Port(rawValue: port)!)
        listener.stateUpdateHandler = { state in
            print("Astur iOS agent listener state: \(state)")
        }
        listener.newConnectionHandler = { [weak self] connection in
            self?.handle(connection)
        }
        listener.start(queue: queue)
        self.listener = listener
    }

    func stop() {
        listener?.cancel()
        listener = nil
    }

    private func handle(_ connection: NWConnection) {
        connection.start(queue: queue)
        read(connection, buffer: Data())
    }

    private func read(_ connection: NWConnection, buffer: Data) {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 64 * 1024) { [weak self] data, _, isComplete, error in
            guard let self else {
                connection.cancel()
                return
            }

            if error != nil {
                connection.cancel()
                return
            }

            var next = buffer
            if let data {
                next.append(data)
            }

            if let request = HttpRequest(data: next) {
                self.respond(to: request, connection: connection)
                return
            }

            if isComplete {
                self.write(
                    connection,
                    status: 400,
                    body: self.errorJson(id: "unknown", code: "BAD_REQUEST", message: "Incomplete Astur HTTP command.")
                )
                return
            }

            self.read(connection, buffer: next)
        }
    }

    private func respond(to request: HttpRequest, connection: NWConnection) {
        do {
            let body: [String: Any]
            if request.method == "GET" && request.path == "/health" {
                let startedAt = Date()
                body = resultJson(
                    agent.dispatch(AsturCommand(id: "health", method: "agent.ping", params: [:])),
                    elapsedMs: elapsedMs(startedAt)
                )
            } else if request.method == "POST" {
                let command = try parseCommand(request.body)
                let startedAt = Date()
                body = resultJson(agent.dispatch(command), elapsedMs: elapsedMs(startedAt))
            } else {
                body = errorJson(
                    id: "unknown",
                    code: "UNSUPPORTED_HTTP_METHOD",
                    message: "Astur agent expects POST commands."
                )
            }

            write(connection, status: 200, body: body)
        } catch {
            write(
                connection,
                status: 500,
                body: errorJson(
                    id: "unknown",
                    code: "SERVER_ERROR",
                    message: String(describing: error)
                )
            )
        }
    }

    private func parseCommand(_ body: Data) throws -> AsturCommand {
        let raw = try JSONSerialization.jsonObject(with: body)
        guard let json = raw as? [String: Any] else {
            throw ServerFailure("Astur command must be a JSON object.")
        }

        let id = (json["id"] as? String)?.nonEmpty ?? "unknown"
        let method = (json["method"] as? String)?.nonEmpty
            ?? (json["command"] as? String)?.nonEmpty

        guard let method else {
            throw ServerFailure("Astur command must include method or command.")
        }

        let params = (json["params"] as? [String: Any])
            ?? (json["payload"] as? [String: Any])
            ?? [:]

        return AsturCommand(id: id, method: method, params: params)
    }

    private func resultJson(_ result: AsturCommandResult, elapsedMs: Int) -> [String: Any] {
        var body: [String: Any] = [
            "id": result.id,
            "ok": result.ok,
            "timing": [
                "totalMs": elapsedMs,
                "agentMs": elapsedMs,
                "hostRoundTrips": 1,
                "usedSnapshot": false,
                "usedLegacyFallback": false
            ]
        ]

        if result.ok {
            body["result"] = result.result ?? NSNull()
            body["data"] = result.result ?? NSNull()
        } else {
            var error: [String: Any] = [
                "code": result.error?.code ?? "UNKNOWN",
                "message": result.error?.message ?? "Unknown iOS agent error."
            ]
            if let details = result.error?.details {
                error["details"] = details
            }
            body["error"] = error
        }

        return body
    }

    private func errorJson(id: String, code: String, message: String) -> [String: Any] {
        [
            "id": id,
            "ok": false,
            "error": [
                "code": code,
                "message": message
            ],
            "timing": [
                "totalMs": 0,
                "agentMs": 0,
                "hostRoundTrips": 1
            ]
        ]
    }

    private func write(_ connection: NWConnection, status: Int, body: [String: Any]) {
        let payload = (try? JSONSerialization.data(withJSONObject: body)) ?? Data()
        var response = Data()
        let header = """
        HTTP/1.1 \(status) \(status == 200 ? "OK" : "Error")\r
        content-type: application/json\r
        content-length: \(payload.count)\r
        connection: close\r
        \r

        """
        response.append(header.data(using: .utf8)!)
        response.append(payload)
        connection.send(content: response, completion: .contentProcessed { _ in
            connection.cancel()
        })
    }

    private func elapsedMs(_ startedAt: Date) -> Int {
        max(0, Int(Date().timeIntervalSince(startedAt) * 1_000))
    }
}

private struct HttpRequest {
    let method: String
    let path: String
    let body: Data

    init?(data: Data) {
        guard let separator = data.range(of: Data("\r\n\r\n".utf8)) else {
            return nil
        }

        let headerData = data[..<separator.lowerBound]
        guard let header = String(data: headerData, encoding: .utf8) else {
            return nil
        }

        let lines = header.components(separatedBy: "\r\n")
        let requestLine = lines.first?.split(separator: " ").map(String.init) ?? []
        guard requestLine.count >= 2 else {
            return nil
        }

        let contentLength = lines
            .first { $0.lowercased().hasPrefix("content-length:") }
            .flatMap { Int($0.split(separator: ":", maxSplits: 1).last?.trimmingCharacters(in: .whitespaces) ?? "") }
            ?? 0

        let bodyStart = separator.upperBound
        guard data.count - bodyStart >= contentLength else {
            return nil
        }

        self.method = requestLine[0]
        self.path = requestLine[1]
        self.body = data[bodyStart..<(bodyStart + contentLength)]
    }
}

private struct ServerFailure: Error, CustomStringConvertible {
    let description: String

    init(_ description: String) {
        self.description = description
    }
}

private extension String {
    var nonEmpty: String? {
        isEmpty ? nil : self
    }
}

File: packages/android-agent/src/androidTest/kotlin/dev/astur/agent/AsturAgentServer.kt
package dev.astur.agent

import android.os.SystemClock
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.net.ServerSocket
import java.net.Socket
import java.net.SocketException
import java.nio.charset.Charset
import java.util.concurrent.Executors
import kotlin.concurrent.thread

class AsturAgentServer(
    private val port: Int,
    private val agent: AsturAgent = AsturAgent()
) {
    private val executor = Executors.newCachedThreadPool()

    @Volatile
    private var running = false
    private var serverSocket: ServerSocket? = null

    fun startBlocking() {
        running = true

        ServerSocket(port).use { server ->
            serverSocket = server

            while (running) {
                try {
                    val client = server.accept()
                    executor.execute {
                        handleClient(client)
                    }
                } catch (error: SocketException) {
                    if (running) {
                        throw error
                    }
                }
            }
        }
    }

    fun startInBackground() {
        thread(name = "AsturAgentServer") {
            startBlocking()
        }
    }

    fun stop() {
        running = false
        runCatching { serverSocket?.close() }
        executor.shutdownNow()
    }

    private fun handleClient(socket: Socket) {
        socket.use { client ->
            try {
                val request = readRequest(client.getInputStream())
                val body = when {
                    request.method == "GET" && request.path == "/health" -> {
                        resultToJson(agent.dispatch(AsturCommand(id = "health", method = "agent.ping")), 0L)
                    }

                    request.method == "POST" -> {
                        val command = parseCommand(request.body)
                        val startedAt = SystemClock.elapsedRealtime()
                        val result = agent.dispatch(command)
                        resultToJson(result, SystemClock.elapsedRealtime() - startedAt)
                    }

                    else -> errorJson(
                        id = "unknown",
                        code = "UNSUPPORTED_HTTP_METHOD",
                        message = "Astur agent expects POST commands."
                    )
                }

                writeJson(client, 200, body)
            } catch (error: Throwable) {
                writeJson(
                    client,
                    500,
                    errorJson(
                        id = "unknown",
                        code = "SERVER_ERROR",
                        message = error.message ?: error.javaClass.name
                    )
                )
            }
        }
    }

    private fun parseCommand(body: String): AsturCommand {
        val json = JSONObject(body)
        val id = json.optString("id").takeIf { it.isNotBlank() } ?: "unknown"
        val method = json.optString("method").takeIf { it.isNotBlank() }
            ?: json.optString("command").takeIf { it.isNotBlank() }
            ?: throw IllegalArgumentException("Astur command must include method or command.")
        val paramsRaw = when {
            json.has("params") -> json.opt("params")
            json.has("payload") -> json.opt("payload")
            else -> JSONObject()
        }

        return AsturCommand(
            id = id,
            method = method,
            params = paramsRaw.asStringMap()
        )
    }

    private fun resultToJson(result: AsturCommandResult, elapsedMs: Long): JSONObject {
        val json = JSONObject()
            .put("id", result.id)
            .put("ok", result.ok)
            .put(
                "timing",
                JSONObject()
                    .put("totalMs", elapsedMs)
                    .put("agentMs", elapsedMs)
                    .put("hostRoundTrips", 1)
                    .put("usedSnapshot", false)
                    .put("usedLegacyFallback", false)
            )

        if (result.ok) {
            json
                .put("result", toJsonValue(result.result))
                .put("data", toJsonValue(result.result))
        } else {
            val details = result.error?.details
            json.put(
                "error",
                JSONObject()
                    .put("code", result.error?.code ?: "UNKNOWN")
                    .put("message", result.error?.message ?: "Unknown Android agent error.")
                    .put("details", toJsonValue(details))
            )

            if (details is Map<*, *>) {
                val diagnostics = details["diagnostics"]
                if (diagnostics != null) {
                    json.put("diagnostics", toJsonValue(diagnostics))
                }
            }
        }

        return json
    }

    private fun errorJson(id: String, code: String, message: String): JSONObject {
        return JSONObject()
            .put("id", id)
            .put("ok", false)
            .put(
                "error",
                JSONObject()
                    .put("code", code)
                    .put("message", message)
            )
            .put(
                "timing",
                JSONObject()
                    .put("totalMs", 0)
                    .put("agentMs", 0)
                    .put("hostRoundTrips", 1)
            )
    }

    private fun readRequest(input: InputStream): HttpRequest {
        val headerBytes = ByteArrayOutputStream()
        var match = 0

        while (true) {
            val next = input.read()
            if (next < 0) {
                throw IllegalArgumentException("Unexpected end of HTTP request headers.")
            }

            headerBytes.write(next)
            match = when {
                match == 0 && next == '\r'.code -> 1
                match == 1 && next == '\n'.code -> 2
                match == 2 && next == '\r'.code -> 3
                match == 3 && next == '\n'.code -> 4
                next == '\r'.code -> 1
                else -> 0
            }

            if (match == 4) {
                break
            }
        }

        val header = headerBytes.toString(UTF_8.name())
        val lines = header.split("\r\n")
        val requestLine = lines.firstOrNull()?.split(" ") ?: emptyList()
        val method = requestLine.getOrNull(0).orEmpty()
        val path = requestLine.getOrNull(1).orEmpty()
        val contentLength = lines
            .firstOrNull { line -> line.lowercase().startsWith("content-length:") }
            ?.substringAfter(":")
            ?.trim()
            ?.toIntOrNull()
            ?: 0
        val body = readBody(input, contentLength).toString(UTF_8)

        return HttpRequest(method = method, path = path, body = body)
    }

    private fun readBody(input: InputStream, contentLength: Int): ByteArray {
        if (contentLength <= 0) {
            return ByteArray(0)
        }

        val body = ByteArray(contentLength)
        var offset = 0

        while (offset < contentLength) {
            val read = input.read(body, offset, contentLength - offset)
            if (read < 0) {
                throw IllegalArgumentException("Unexpected end of HTTP request body.")
            }

            offset += read
        }

        return body
    }

    private fun writeJson(socket: Socket, status: Int, body: JSONObject) {
        val payload = body.toString().toByteArray(UTF_8)
        val statusText = if (status == 200) "OK" else "Error"
        val header = buildString {
            append("HTTP/1.1 ")
            append(status)
            append(' ')
            append(statusText)
            append("\r\ncontent-type: application/json\r\ncontent-length: ")
            append(payload.size)
            append("\r\nconnection: close\r\n\r\n")
        }.toByteArray(UTF_8)

        socket.getOutputStream().apply {
            write(header)
            write(payload)
            flush()
        }
    }

    private fun Any?.asStringMap(): Map<String, Any?> {
        if (this == null || this == JSONObject.NULL) {
            return emptyMap()
        }

        if (this !is JSONObject) {
            throw IllegalArgumentException("Astur command params/payload must be an object.")
        }

        val result = linkedMapOf<String, Any?>()
        val keys = keys()

        while (keys.hasNext()) {
            val key = keys.next()
            result[key] = fromJsonValue(opt(key))
        }

        return result
    }

    private fun fromJsonValue(value: Any?): Any? {
        return when (value) {
            null, JSONObject.NULL -> null
            is JSONObject -> value.asStringMap()
            is JSONArray -> {
                val list = mutableListOf<Any?>()
                for (index in 0 until value.length()) {
                    list += fromJsonValue(value.opt(index))
                }
                list
            }

            else -> value
        }
    }

    private fun toJsonValue(value: Any?): Any {
        return when (value) {
            null -> JSONObject.NULL
            is Map<*, *> -> {
                val json = JSONObject()
                for ((key, child) in value) {
                    if (key != null) {
                        json.put(key.toString(), toJsonValue(child))
                    }
                }
                json
            }

            is Iterable<*> -> {
                val json = JSONArray()
                for (child in value) {
                    json.put(toJsonValue(child))
                }
                json
            }

            is Array<*> -> {
                val json = JSONArray()
                for (child in value) {
                    json.put(toJsonValue(child))
                }
                json
            }

            else -> value
        }
    }

    private data class HttpRequest(
        val method: String,
        val path: String,
        val body: String
    )

    private companion object {
        private val UTF_8: Charset = Charsets.UTF_8
    }
}
